tools/nodejs_api/src_nodejs/connection.js

```diff
@@ -4,17 +4,44 @@ const QueryResult = require("./queryResult.js");
 class Connection {
     #connection;
     #database;
+    #initPromise;
+    #initialized;
     constructor(database, numThreads = 0) {
-        if (typeof database !== "object" || database.constructor.name !==  "Database" || typeof numThreads !== "number" || !Number.isInteger(numThreads)){
+        if (typeof database !== "object" || database.constructor.name !== "Database" || typeof numThreads !== "number" || !Number.isInteger(numThreads)) {
             throw new Error("Connection constructor requires a database object and optional numThreads integer as argument(s)");
         }
         this.#database = database;
         this.#connection = new kuzu.NodeConnection(database.database, numThreads);
+        this.#initPromise = this.#connection.getConnection();
+        this.#initialized = false;
     }
 
-    execute(query, opts = {}) {
+    async getConnection() {
+        if (this.#initialized) {
+            return this.#connection;
+        }
+        const promiseResult = await this.#initPromise;
+        if (promiseResult) {
+            this.#connection.transferConnection();
+            this.#initialized = true;
+        } else {
+            throw new Error("GetConnection Failed");
+        }
+        return this.#connection;
+    }
+
+    async execute(query, opts = {}) {
+        const connection = await this.getConnection();
+        const optsKeys = Object.keys(opts);
         if (typeof opts !== "object") {
             throw new Error("optional opts in execute must be an object");
+        } else if (optsKeys.length > 2) {
```

Comment:
No need to do this type of check in Javascript. The idea of [Duck-typing](https://medium.com/@eamonocallaghan/what-is-duck-typing-in-javascript-f3eb10853361) is common in Javascript. So you should proceed as long as you get what you need. If there is additional vars / functions, you can just ignore them instead of raising an error.